### PR TITLE
Metric SDK: Sum duplicate async observations regardless of filtering

### PR DESCRIPTION
**Pull request by dashpole**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4217
It addresses this issue by (always) summing duplicate observations regardless of whether or not a filter is applied.  This simplifies the implementation as well by removing special code for the filtered vs non-filtered case.

### CODE REVIEW

File: sdk/metric/internal/aggregate/sum.go [hunk 1]
```diff
@@ -269,24 +214,19 @@ func (s *precomputedDeltaSum[N]) Aggregation() metricdata.Aggregation {
...
-		// TODO (#3006): This will use an unbounded amount of memory if there
-		// are unbounded number of attribute sets being aggregated. Attribute
-		// sets that become "stale" need to be forgotten so this will not
-		// overload the system.
+		newReported[attr] = value
+		// Unused attribute sets do not report.
+		delete(s.values, attr)
 	}
+	// Unused attribute sets are forgotten.
+	s.reported = newReported
```
Comment (dashpole) [inline sdk/metric/internal/aggregate/sum.go:229]: I would like feedback on this change in particular.  It only keeps track of the previous collection interval's values for the purpose of computing deltas in subsequent collection intervals. 
This has the benefit of ensuring that "churning" attribute sets don't cause a memory leak in precomputed delta sums.  Without the change, we would continue to track the previous value for an attribute set forever.

File: sdk/metric/internal/aggregate/sum_test.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ func testSum[N int64 | float64](t *testing.T) {
 		MeasurementN: defaultMeasurements,
 		CycleN:       defaultCycles,
 	}
+	totalMeasurements := defaultGoroutines * defaultMeasurements
```
Comment (dashpole) [inline sdk/metric/internal/aggregate/sum_test.go:40]: This test produces a _ton_ of duplicate measurements within a single Aggregation.  Since these now sum (instead of keeping the last value), I scaled the expected value of precomputed tests by the number of measurements that go into them.

**Comment by dashpole** [comment]
I'll wait to request review until I can rebase this on https://github.com/open-telemetry/opentelemetry-go/pull/4290

**Comment by dashpole** [comment]
Rebased, and ready for review!

File: sdk/metric/internal/aggregate/sum_test.go [hunk 3]
```diff
@@ -167,118 +168,92 @@ func TestDeltaSumReset(t *testing.T) {
...
 
 	// Filtered values should not persist.
 	agg.Aggregate(5, attrs)
-	got = agg.Aggregation()
-	// measured(+): 5, previous(-): 18, filtered(+): 0
-	want.DataPoints = []metricdata.DataPoint[int64]{point[int64](attrs, -13)}
-	metricdatatest.AssertAggregationsEqual(t, want, got, opt)
+	// measured(+): 5, previous(-): 20
+	want.DataPoints = []metricdata.DataPoint[int64]{point[int64](attrs, -15)}
+	metricdatatest.AssertAggregationsEqual(t, want, agg.Aggregation(), opt)
```
Comment (MrAlias) [inline sdk/metric/internal/aggregate/sum_test.go:205]: There is no distinction between filtered and non-filtered at this point. It seems like this section could just be removed as it will likely cause future confusion.

File: sdk/metric/internal/aggregate/sum_test.go [hunk 4]
```diff
@@ -167,118 +168,92 @@ func TestDeltaSumReset(t *testing.T) {
...
 	want.DataPoints = []metricdata.DataPoint[int64]{point[int64](attrs, 15)}
-	metricdatatest.AssertAggregationsEqual(t, want, got, opt)
+	metricdatatest.AssertAggregationsEqual(t, want, agg.Aggregation(), opt)
 	agg.Aggregate(7, attrs)
-	got = agg.Aggregation()
-	// measured(+): 7, previous(-): 20, filtered(+): 0
+	// measured(+): 7, previous(-): 20
 	want.DataPoints = []metricdata.DataPoint[int64]{point[int64](attrs, -13)}
-	metricdatatest.AssertAggregationsEqual(t, want, got, opt)
+	metricdatatest.AssertAggregationsEqual(t, want, agg.Aggregation(), opt)
```
Comment (MrAlias) [inline sdk/metric/internal/aggregate/sum_test.go:218]: Similar here; this could just be removed.

File: sdk/metric/internal/aggregate/sum_test.go [hunk 5]
```diff
@@ -167,118 +168,92 @@ func TestDeltaSumReset(t *testing.T) {
...
-	agg.(precomputeAggregator[int64]).aggregateFiltered(3, attrs)
+	// Duplicates should add.
+	agg.Aggregate(3, attrs)
 	agg.Aggregate(7, attrs)
-	agg.(precomputeAggregator[int64]).aggregateFiltered(10, attrs)
-	got = agg.Aggregation()
+	agg.Aggregate(10, attrs)
 	want.DataPoints = []metricdata.DataPoint[int64]{point[int64](attrs, 20)}
-	metricdatatest.AssertAggregationsEqual(t, want, got, opt)
+	metricdatatest.AssertAggregationsEqual(t, want, agg.Aggregation(), opt)
```
Comment (MrAlias) [inline sdk/metric/internal/aggregate/sum_test.go:256]: same

File: sdk/metric/meter_test.go [hunk 6]
```diff
@@ -960,9 +960,6 @@ func TestGlobalInstRegisterCallback(t *testing.T) {
 	_, err = preMtr.RegisterCallback(cb, preInt64Ctr, preFloat64Ctr, postInt64Ctr, postFloat64Ctr)
 	assert.NoError(t, err)
 
-	_, err = preMtr.RegisterCallback(cb, preInt64Ctr, preFloat64Ctr, postInt64Ctr, postFloat64Ctr)
-	assert.NoError(t, err)
-
```
Comment (MrAlias) [inline sdk/metric/meter_test.go:965]: Is this related to the changes, or just clean up?

I'm fine either way, just wondering if I understand this.

Comment (dashpole) [reply sdk/metric/meter_test.go:965]: Just cleanup, but revealed by the change.  The callback was registered twice (by accident?), which now causes values to be 2x.

File: sdk/metric/internal/aggregate/sum_test.go (hunk 5 above)
Comment (dashpole) [reply sdk/metric/internal/aggregate/sum_test.go]: removed

File: sdk/metric/internal/aggregate/sum_test.go (hunk 4 above)
Comment (dashpole) [reply sdk/metric/internal/aggregate/sum_test.go]: removed

File: sdk/metric/internal/aggregate/sum_test.go (hunk 3 above)
Comment (dashpole) [reply sdk/metric/internal/aggregate/sum_test.go]: removed

**Review by MrAlias — APPROVED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -29,6 +29,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   The `AttributeKeys` fields allows users to specify an allow-list of attributes allowed to be recorded for a view.
   This change is made to ensure compatibility with the OpenTelemetry specification. (#4288)
 - If an attribute set is omitted from an async callback, the previous value will no longer be exported. (#4290)
+- If an attribute set is Observed multiple times in an async callback, the values will be summed instead of the last observation winning. (#4289)
```
Comment (pellared) [inline CHANGELOG.md:32]:
```suggestion
- If an attribute set is observed multiple times in an async callback, the values will be summed instead of the last observation winning. (#4289)
```

**Review by pellared — APPROVED** [review]

**Comment by MrAlias** [comment]
@dashpole it looks like https://github.com/open-telemetry/opentelemetry-go/pull/4304 was merged prior to this. Can you rebase this (sorry about the conflicts)?

@open-telemetry/go-approvers PTAL at this.

**Comment by dashpole** [comment]
no worries.  Rebased
